src/main/java/us/kbase/userprofile/MongoController.java

```diff
@@ -195,89 +210,95 @@ public UserProfile getProfile(String username) {
+				Document replacement = new Document("user", user);
+				if(up.getProfile() != null) {
+					if(up.getProfile().asJsonNode().isObject()) {
+						replacement.put("profile", Document.parse(up.getProfile().asJsonNode().toString()));
+					} else {
 						throw new RuntimeException("Profile must be an object if defined.");
 					}
 				} else {
 					replacement.put("profile", null);
 				}
 				System.out.println(replacement);
-				profiles.update(
-						new BasicDBObject("user.username",up.getUser().getUsername()),
-						new BasicDBObject("$set",replacement));
+				profiles.updateOne(
+						new Document("user.username", up.getUser().getUsername()),
+						new Document("$set", replacement));
 			} else {
-				DBObject user = new BasicDBObject("username",up.getUser().getUsername())
+				Document user = new Document("username", up.getUser().getUsername())
 									.append("realname", up.getUser().getRealname())
 									.append("thumbnail", up.getUser().getThumbnail());
-				
-				DBObject profile = new BasicDBObject("user",user);
-				if(up.getProfile()!=null) {
-					if(up.getProfile().asJsonNode().isObject())
-						profile.put("profile", JSON.parse(up.getProfile().asJsonNode().toString()));
-					else {
+
+				Document profile = new Document("user", user);
+				if (up.getProfile() != null) {
+					if (up.getProfile().asJsonNode().isObject()) {
+						profile.put("profile", Document.parse(up.getProfile().asJsonNode().toString()));
+					} else {
 						throw new RuntimeException("Profile must be an object if defined.");
 					}
 				} else {
 					profile.put("profile", null);
 				}
-				profiles.insert(profile);
+				profiles.insertOne(profile);
 			}
 		}
 	
 	
 	// assume UserProfile is validated
 	public void updateProfile(UserProfile up) {
 		if(exists(up.getUser().getUsername())) {
-			DBObject user = new BasicDBObject("username",up.getUser().getUsername());
-			if(up.getUser().getRealname()!=null)
+			Document user = new Document("username", up.getUser().getUsername());
+
+			if(up.getUser().getRealname() != null) {
 				user.put("realname", up.getUser().getRealname());
-			if(up.getUser().getThumbnail()!=null)
+			}
+			if(up.getUser().getThumbnail() != null) {
 				user.put("thumbnail", up.getUser().getThumbnail());
+			}
 			
-			DBObject update = new BasicDBObject("user",user);
-			if(up.getProfile()!=null) {
+			Document update = new Document("user", user);
+			if(up.getProfile() != null) {
 				JsonNode profileNode = up.getProfile().asJsonNode();
 				System.out.println(profileNode);
-				if( profileNode.isObject() ) {
-					Iterator<Entry<String, JsonNode>> fields = profileNode.fields();
-					while(fields.hasNext()) {
-						Entry<String,JsonNode> e= fields.next();
-						update.put("profile."+e.getKey(), JSON.parse(e.getValue().toString()));
-					}
+
+				if(profileNode.isObject()) {
+					Document mergedProfile = createMergedProfile(up.getUser().getUsername(), profileNode);
+					update.put("profile", mergedProfile);
```

Comment:
I'm confused by this code. It looks like there's no way `update` has a profile entry, and so merging is just effectively copying in the profile from `up`.

In contrast the `createMergedProfile` method actually pulls the user and merges the profiles, changing the behavior.

 Am I missing something?

Reply:
> It looks like there's no way update has a profile entry

Why?

Reply:
Because it's created without one.

Reply:
It is confusing. 

The code only updates fields that are explicitly included in the `up`. Fields not mentioned remain unchanged.

https://github.com/kbase/user_profile/blob/develop/src/test/java/us/kbase/test/userprofile/FullServerTest.java#L130-L131

Reply:
Right, so in the old code you could literally just copy the profile field from `up` directly into `update` and have the same effect. That why I'm wondering why the new code has different behavior and if it should have that behavior

Reply:
I tried the approach that the old code used, specifically `update.put("profile." + e.getKey(), ...)`, but it did not work in the MongoDB 7 context. So, I converted it to a merge.

Reply:
It’s kind of mixed up here. I would personally rewrite the entire function differently in the MongoDB 7 context and might use something like this:

```            // Update the document in place
            collection.updateOne(
                Filters.eq("username", username), // Filter by username
                Updates.combine(
                    Updates.set("fieldName1", newField1Value), // Update first field
                    Updates.set("fieldName2", newField2Value)  // Update second field
                )
            );
```

Reply:
ah, the current code looks up a particular document based on the username three times.

Reply:
I mean the old code completely discarded the prior profile as far as I can tell. It just creates the update document and moves the profile from `up` into it, and so any preexisting profile is lost when `update` is sent to mongo.

Reply:
> I mean the old code completely discarded the prior profile as far as I can tell.

Preexisting profile fields are either retained or overwritten, but never lost.

Reply:
Oh, because `update` is being `$set`, I see.

In that case you definitely don't want to pull the old fields to the server, update, and then push the whole thing back, that's a classic race condition. You'll want to do it with a `$set` as well. That's used all over the place in the workspace and auth which are running on mongo 7 in CI so it should work here

Reply:
It works now, but I'm not 100% sure why asText() works while toString() does not.

Reply:
https://fasterxml.github.io/jackson-databind/javadoc/2.7/com/fasterxml/jackson/databind/JsonNode.html#asText()
vs
https://fasterxml.github.io/jackson-databind/javadoc/2.7/com/fasterxml/jackson/databind/JsonNode.html#toString()

We shouldn't be using JsonNodes at all here, but regular Java types. That's for later though

Reply:
And actually - is it guaranteed the field is always text? I'm not sure that 's true, and they may be why the code is doing the insane serialize -> deserialize hack

Reply:
> We shouldn't be using JsonNodes at all here, but regular Java types. That's for later though

How would you do it exactly?

Reply:
so the issue is that UserProfile is an UnspecifiedObject which means it can be literally anything, which is a design mistake IMO
https://github.com/kbase/user_profile/blob/develop/UserProfile.spec#L20

If we could make that a mapping<string, UnspecifiedObject> it's make things easier to deal with.

That being said, if you're just trying to translate from API to DB, then using Object as an intermediary would be the safe way to go, so something like unspecifiedObject.asClassInstance(Object.class) and document.get("profile", Object.class).

For the merge case, use unspecifiedObject.isMap() and unspecifiedObject.asMap() to run through the keys, and for each value transform it to an Object as above.

Haven't looked super closely at the code or tested any of this obviously

Reply:
You've covered the case where the field is a map (Object in `JSONNode` speak) but what if it's a list or an number? What the code should do is just take the value in the `JSONNode`, whatever it is, and put it into the `Document`.

Reply:
Document.parse(e.getValue()) did not work, though. I am not sure if there is a unified way to auto-parse complex fields in the Mongo7 context, as there is in Mongo3. That's why I tend to handle them by type.

Reply:
That parse line is expecting a JSON string right? e.getValue() returns a JsonNode so that won't work. You'll need to turn that JsonNode into a java Object since it can be a scalar or a container

Reply:
Document.parse() expects a JSON string. After I convert that JsonNode into a Java Object, I still need to convert it back to a string. It did not work.

```
Object val = e.getValue();
update.put("profile." + e.getKey(), Document.parse(val.toString()));
```

Reply:
Assuming I've got this right, you don't use a json string, you just convert the jsonnode to a regular old java object and then insert that as the value in put()

Reply:
Like this?

```
Object val = e.getValue();
update.put("profile." + e.getKey(), val);
```

Reply:
No, you have to actually convert the JsonNode to an object. That is effectively just casting it up the inheritance hierarchy, it's still the JsonNode type. I'm pretty sure the Mongo client doesn't know how to convert those to BSON

Reply:
It works now. I assume I need to add a test to cover missing lines/cases for thrown exceptions?

Reply:
Using this method might make things simpler: https://github.com/kbase/java_common/blob/develop/src/main/java/us/kbase/common/service/UObject.java#L465

Is there any way you can set up a JsonNode so that Jackson can't parse it?

Reply:
It is simpler because it is caught inside UObject.transformJacksonToObject?
Now I can remove the try-catch clause from outside.

Reply:
And you don't need the mapper

Reply:
A follow-up question: checked exceptions must be either caught or declared. When do you decide to use one over the other in Java?

Reply:
I catch and rethrow a checked exception as a runtime exception when
* It's impossible for the checked exception to occur
* There's nothing that can be done about the checked exception other than to crash the program or fail the request

You'll see there's a lot of places in java_common that don't follow those rules. I think I'm a little idiosyncratic in that I like checked exceptions because they force the programmer to be explicit about what errors a function may throw.

I use checked exceptions when the programmer or user might be able to do something useful with the information in the exception. For instance, in my services checked exceptions are always 4XX errors vs. unchecked which are 500s.

Reply:
That's very insightful. I am more interested in when to use try-catch for a checked exception versus when to declare a checked exception.

Reply:
That's the question I was answering. One case I missed would be if you want to rethrow as a different type of exception, then you'd catch and rethrow, but otherwise the above should cover it.

Obviously you'd catch if you want to run some code block if that exception occurs

Reply:
Thanks for clarifying! 

I believe this PR is ready to merge now. Let me know if there's anything else you'd like to discuss.